File: apps/SemanticGraph.Visualizer/src/SemanticGraph.Visualizer.Client/Models/GraphNodeDetailsDto.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace SemanticGraph.Visualizer.Client.Models;

public sealed record GraphNodeDetailsDto(
    [property: JsonPropertyName("nodeId")] string NodeId,
    [property: JsonPropertyName("kind")] string Kind,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("displayLabel")] string DisplayLabel,
    [property: JsonPropertyName("qualifiedName")] string? QualifiedName,
    [property: JsonPropertyName("language")] string Language,
    [property: JsonPropertyName("sourceFilePath")] string? SourceFilePath,
    [property: JsonPropertyName("startLine")] long? StartLine,
    [property: JsonPropertyName("startCol")] long? StartCol,
    [property: JsonPropertyName("endLine")] long? EndLine,
    [property: JsonPropertyName("endCol")] long? EndCol,
    [property: JsonPropertyName("selectionStartLine")] long? SelectionStartLine,
    [property: JsonPropertyName("selectionStartCol")] long? SelectionStartCol,
    [property: JsonPropertyName("containerNodeId")] string? ContainerNodeId,
    [property: JsonPropertyName("containerDisplayLabel")] string? ContainerDisplayLabel,
    [property: JsonPropertyName("firstSeenRunId")] long? FirstSeenRunId,
    [property: JsonPropertyName("lastSeenRunId")] long? LastSeenRunId,
    [property: JsonPropertyName("propertiesJson")] JsonElement PropertiesJson,
    [property: JsonPropertyName("incomingEdgeCount")] long IncomingEdgeCount,
    [property: JsonPropertyName("outgoingEdgeCount")] long OutgoingEdgeCount,
    [property: JsonPropertyName("relations")] IReadOnlyList<GraphNodeRelationSummaryDto> Relations,
    [property: JsonPropertyName("occurrences")] IReadOnlyList<GraphNodeOccurrenceDto> Occurrences);

File: crates/semantic-graph-visualizer-server/src/dto/graph_node_details_dto.rs
use crate::dto::{GraphNodeOccurrenceDto, GraphNodeRelationSummaryDto};

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct GraphNodeDetailsDto {
    pub node_id: String,
    pub kind: String,
    pub name: String,
    pub display_label: String,
    pub qualified_name: Option<String>,
    pub language: String,
    pub source_file_path: Option<String>,
    pub start_line: Option<i64>,
    pub start_col: Option<i64>,
    pub end_line: Option<i64>,
    pub end_col: Option<i64>,
    pub selection_start_line: Option<i64>,
    pub selection_start_col: Option<i64>,
    pub container_node_id: Option<String>,
    pub container_display_label: Option<String>,
    pub first_seen_run_id: Option<i64>,
    pub last_seen_run_id: Option<i64>,
    pub properties_json: Value,
    pub incoming_edge_count: i64,
    pub outgoing_edge_count: i64,
    pub relations: Vec<GraphNodeRelationSummaryDto>,
    pub occurrences: Vec<GraphNodeOccurrenceDto>,
}
